math/inv_sqrt.cpp

```diff
@@ -0,0 +1,100 @@
+/**
+ * @file
+ * @brief Implementation of [the inverse square root
+ * Root](https://medium.com/hard-mode/the-legendary-fast-inverse-square-root-e51fee3b49d9).
+ * @details
+ * Two implementation to calculate inverse inverse root,
+ * from Quake III Arena (C++ version) and with a standard library (`cmath`).
+ * This algorithm is used to calculate shadows in Quake III Arena.
+ */
+
+#include <cassert>   /// for assert
+#include <cmath>     /// for `std::sqrt`
+#include <iostream>  /// for IO operations
+#include <limits>    /// for numeric_limits
+
+/**
+ * @brief This is the function that calculates the fast inverse square root.
+ * The following code is the fast inverse square root implementation from
+ * Quake III Arena (Adapted for C++). More information can be found at
+ * [Wikipedia](https://en.wikipedia.org/wiki/Fast_inverse_square_root)
+ * @tparam T floating type
+ * @tparam iterations inverse square root, the greater the number of
+ * iterations, the more exact the result will be (1 or 2).
+ * @param x value to calculate
+ * @return the inverse square root
+ */
+template <typename T = double, char iterations = 2>
+inline T Fast_InvSqrt(T x) {
+    using Tint = typename std::conditional<sizeof(T) == 8, std::int64_t,
+                                           std::int32_t>::type;
+    T y = x;
+    T x2 = y * 0.5;
+    Tint i = *(Tint *)&y;
+    i = (sizeof(T) == 8 ? 0x5fe6eb50c7b537a9 : 0x5f3759df) - (i >> 1);
+    y = *(T *)&i;
+
+    y = *reinterpret_cast<T *>(&i);
+
+    y = y * (1.5 - (x2 * y * y));
+    if (iterations == 2) {
+        y = y * (1.5 - (x2 * y * y));
+    }
+    return y;
```

Comment:
Please add documentation of what this code does here.

Reply:
> Please add documentation of what this code does here.

Comments have been added